File: backend/schemas/insurance.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field, field_serializer


class InsuranceCreate(BaseModel):
    provider_name: str = Field(..., min_length=1, max_length=200)
    policy_number: Optional[str] = None
    policy_type: Optional[str] = None
    portal_url: Optional[str] = None
    coverage_amount: Optional[Decimal] = None
    deductible: Optional[Decimal] = None
    annual_premium: Optional[Decimal] = None
    effective_date: Optional[date] = None
    expiration_date: Optional[date] = None
    renewal_date: Optional[date] = None
    agent_name: Optional[str] = None
    agent_phone: Optional[str] = None
    agent_email: Optional[str] = None


class InsuranceUpdate(BaseModel):
    provider_name: Optional[str] = None
    policy_number: Optional[str] = None
    policy_type: Optional[str] = None
    portal_url: Optional[str] = None
    coverage_amount: Optional[Decimal] = None
    deductible: Optional[Decimal] = None
    annual_premium: Optional[Decimal] = None
    effective_date: Optional[date] = None
    expiration_date: Optional[date] = None
    renewal_date: Optional[date] = None
    agent_name: Optional[str] = None
    agent_phone: Optional[str] = None
    agent_email: Optional[str] = None


class InsuranceResponse(BaseModel):
    id: uuid.UUID
    property_id: uuid.UUID
    provider_name: str
    policy_number: Optional[str] = None
    policy_type: Optional[str] = None
    portal_url: Optional[str] = None
    coverage_amount: Optional[Decimal] = None
    deductible: Optional[Decimal] = None
    annual_premium: Optional[Decimal] = None
    effective_date: Optional[date] = None
    expiration_date: Optional[date] = None
    renewal_date: Optional[date] = None
    agent_name: Optional[str] = None
    agent_phone: Optional[str] = None
    agent_email: Optional[str] = None
    is_active: bool = True
    ended_at: Optional[date] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @field_serializer("id", "property_id")
    def serialize_uuids(self, value: uuid.UUID) -> str:
        return str(value)

File: lib/api/insurance.ts
import { api } from "./client";

export interface InsuranceData {
  id: string;
  property_id: string;
  provider_name: string;
  policy_number?: string | null;
  policy_type?: string | null;
  portal_url?: string | null;
  coverage_amount?: number | null;
  deductible?: number | null;
  annual_premium?: number | null;
  effective_date?: string | null;
  expiration_date?: string | null;
  renewal_date?: string | null;
  agent_name?: string | null;
  agent_phone?: string | null;
  agent_email?: string | null;
  is_active: boolean;
  ended_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface InsuranceCreateData {
  provider_name: string;
  policy_number?: string;
  policy_type?: string;
  portal_url?: string;
  coverage_amount?: number | null;
  deductible?: number | null;
  annual_premium?: number | null;
  effective_date?: string | null;
  expiration_date?: string | null;
  renewal_date?: string | null;
  agent_name?: string | null;
  agent_phone?: string | null;
  agent_email?: string | null;
}

export function getActivePolicy(propertyId: string): Promise<InsuranceData | null> {
  return api.get<InsuranceData | null>(`/properties/${propertyId}/insurance`);
}

export function createPolicy(propertyId: string, data: InsuranceCreateData): Promise<InsuranceData> {
  return api.post<InsuranceData>(`/properties/${propertyId}/insurance`, data);
}

export function updatePolicy(policyId: string, data: Partial<InsuranceCreateData>): Promise<InsuranceData> {
  return api.patch<InsuranceData>(`/insurance/${policyId}`, data);
}

export function deletePolicy(policyId: string): Promise<void> {
  return api.delete<void>(`/insurance/${policyId}`);
}
